core/v2/assignment.go

```diff
@@ -2,72 +2,191 @@ package v2
 		}
 
-		found = true
-		stakePercentage := new(big.Float).Quo(
-			new(big.Float).SetInt(stake.Stake),
-			new(big.Float).SetInt(total.Stake),
-		)
-		if maxStakePercentage.Cmp(stakePercentage) < 0 {
-			maxStakePercentage = stakePercentage
+	}
+
+	return assignments, nil
+}
+
+func AddAssignmentsForQuorum(assignments map[core.OperatorID]*Assignment, state *core.OperatorState, blobParams *core.BlobVersionParameters, quorum core.QuorumID) (map[core.OperatorID]*Assignment, error) {
+
+	dummyAssignments, err := GetAssignmentsForQuorum(state, blobParams, quorum)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get assignments for quorum %d: %w", quorum, err)
+	}
+
+	orderedOps, _, err := getOrderedOperators(state, quorum)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get ordered operators for quorum %d: %w", quorum, err)
+	}
+
+	usedIndices := make(map[uint32]struct{})
+
+	newAssignments := make(map[core.OperatorID]*Assignment)
+
+	for _, id := range orderedOps {
+		newAssignmentIndicesCount := len(dummyAssignments[id].Indices)
+
+		if newAssignmentIndicesCount > len(assignments[id].Indices) {
+			newAssignmentIndicesCount = len(assignments[id].Indices)
+		}
+
+		newAssignments[id] = &Assignment{
+			Indices: assignments[id].Indices[:newAssignmentIndicesCount],
+		}
+
+		if newAssignmentIndicesCount < len(assignments[id].Indices) {
+			excessIndices := assignments[id].Indices[newAssignmentIndicesCount:]
+			for _, index := range excessIndices {
+				usedIndices[index] = struct{}{}
+			}
 		}
 	}
 
-	if !found {
-		return nil, ErrNotFound
+	availableIndices := make([]uint32, 0)
+	for i := uint32(0); i < blobParams.NumChunks; i++ {
+		if _, ok := usedIndices[i]; !ok {
+			availableIndices = append(availableIndices, i)
+		}
 	}
 
-	return maxStakePercentage, nil
-}
+	for _, id := range orderedOps {
 
-// getRelevantOperators gets all of the operators that are relevant to the blob
-func getRelevantOperators(state *core.OperatorState, quorums []core.QuorumID) (map[core.OperatorID]struct{}, error) {
-	if state == nil {
-		return nil, fmt.Errorf("state cannot be nil")
+		newAssignmentIndicesCount := len(dummyAssignments[id].Indices)
+		if newAssignmentIndicesCount > len(newAssignments[id].Indices) {
+
+			// Add available indices to new assignments
+			newAssignments[id].Indices = append(newAssignments[id].Indices, availableIndices[:newAssignmentIndicesCount]...)
+
+			// Remove used indices from available indices
+			availableIndices = availableIndices[newAssignmentIndicesCount:]
+		}
 	}
 
-	operators := make(map[core.OperatorID]struct{}, 0)
-	for _, q := range quorums {
-		ops, ok := state.Operators[q]
-		if !ok || len(ops) == 0 {
-			return nil, fmt.Errorf("no operators found for quorum %d", q)
+	return newAssignments, nil
+}
+
+func MergeAssignmentsAndCap(assignments []map[core.OperatorID]*Assignment, blobParams *core.BlobVersionParameters) map[core.OperatorID]Assignment {
+
+	_mergedAssignments := make(map[core.OperatorID]*Assignment)
```

Comment:
nit, would it be possible to rename this to something more descriptive? Not a fan of having both `_mergedAssignments` and `mergedAssignments` as variables. Perhaps you can do `mergedAssignments` -> `finalMergedAssignments`, and `_mergedAssignments` -> `mergedAssignments`.